sdk/core/core-rest-pipeline/src/httpHeaders.ts

```diff
@@ -59,8 +72,8 @@ class HttpHeadersImpl implements HttpHeaders {
    */
   public toJSON(): RawHttpHeaders {
     const result: RawHttpHeaders = {};
-    for (const [key, value] of this._headersMap) {
-      result[key] = value;
+    for (const entry of this._headersMap.values()) {
+      result[entry.name] = entry.value;
```

Comment:
@deyaaeldeen [mentioned that ](https://github.com/Azure/azure-sdk-for-js/pull/18321#pullrequestreview-787336529) this change of behavior could break existing code that assumes lower case key
 > The [LRO engine assumes the retry header has lower case](https://github.com/Azure/azure-sdk-for-js/blob/58beb3e7d18650c5d961c5099e2519f58a963d99/sdk/core/core-lro/src/lroEngine/stateMachine.ts#L65) so perhaps we will need to explicitly lower the case there?

yeah I think it's better to fix the incorrect assumption.

Reply:
Updating core-lro sounds good to me.

Reply:
@deyaaeldeen I looked into this a bit more and would like your feedback.

Since the public surface of `core-lro` declares that `RawResponse.headers` are lowercase, perhaps we should preserve this behavior?

https://github.com/Azure/azure-sdk-for-js/blob/58beb3e7d18650c5d961c5099e2519f58a963d99/sdk/core/core-lro/src/lroEngine/models.ts#L67

The downside is that `core-lro` doesn't actually produce `RawResponse` objects, it only consumes them via generated code. So this is tricky to fix. I see some options:

1. We change `core-lro` wherever it reads headers to look for a lowercased version of each property name (kinda messy)
2. We update the codegen to lowercase all header names before passing them into `core-lro`

For the second one, I could add an optional parameter to `toJSON()` that lowercases the header names, to make the codegen very similar with `toJSON({lowercase: true})` being the only change.

What do you think? Should we change the contract or update codegen?

Reply:
@xirzec I think the main trade-off here is that normalization makes it easy to lookup standard headers but makes looking up custom headers inconvenient because these lookups are usually case-sensitive.

I think the best way to handle this is for core to provide two APIs, one for each case. This way, the generated code does not need to be updated.

Reply:
@deyaaeldeen that's an interesting idea. Maybe we go the other way and have `toJSON({preserveCase: true})` for this? @jeremymeng would that work for Storage?

Reply:
@xirzec Storage doesn't use the headers directly.  In core if we send the request using the original casing the it should be fine

https://github.com/Azure/azure-sdk-for-js/blob/61331e67733c1a67a34df61f3050d920af1bd472/sdk/core/core-rest-pipeline/src/nodeHttpClient.ts#L211

Reply:
OK I ended up splitting the difference. I kept the original casing when iterating over the collection, since then you're not hunting for a particular value (and you can easily toLowercase each thing as you iterate), but I switched `toJSON` back to lowercasing unless a flag is passed in. 

Now xhrHttpClient and nodeHttpClient should both do the right thing, which is send original case for header names. @deyaaeldeen @jeremymeng please take another look when you have time 😄